Add Dockerfile with docker-compose orchestration, edit grasa_event_locator settings

nlMeminger opened the pull request:
Added docker files for docker based dev testing and fixed some things to make the template project actually work.

Commit:
temporary commit

Signed-off-by: Nathaniel <nathaniel@abba.larrimore.de>

Commit:
Adding mysql container

Commit:
adding docker-compose.yml and adding ignore rule for mysql/

justwheel commented on Dockerfile:
```diff
@@ -0,0 +1,18 @@
+                                                                                 
+RUN apt-get update \                                                             
+    && apt-get install -y --no-install-recommends \                              
+        postgresql-client \                                                      
```
Just curious, why are we installing the PostgreSQL client package?

justwheel replied:
> **@nlMeminger**:
> The dangers of finding guides on the internet and copying and pasting.
> Realistically it shouldn’t be there, but it doesn’t hurt right now.

Okay, cool. I wanted to double-check to be sure. In this case, you can drop the `apt-get install` step completely to reduce container build time.

justwheel committed:
Add --system flag to pipenv install step

This commit adds the `--system` flag to `pipenv install`. This ensures packages are installed using system paths instead of installing into a virtual environment.

nlMeminger commented:
The dangers of finding guides on the internet and copying and pasting.
Realistically it shouldn’t be there, but it doesn’t hurt right now.

On Sun, Sep 22, 2019 at 17:18 Justin W. Flory <notifications@github.com>
wrote:

> *@jwflory* commented on this pull request.
> ------------------------------
>
> In Dockerfile
> <https://github.com/jwflory/django-rit-grasa/pull/11#discussion_r326920105>
> :
>
> > @@ -0,0 +1,18 @@
> +FROM python:3.6
> +
> +RUN apt-get update \
> +    && apt-get install -y --no-install-recommends \
> +        postgresql-client \
>
> Just curious, why are we installing the PostgreSQL client package?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jwflory/django-rit-grasa/pull/11?email_source=notifications&email_token=AELFBEN3ANONPGIHDYXYUP3QK7OL7A5CNFSM4IY7DBT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFQFZVY#pullrequestreview-291527895>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AELFBEOCDANS36C6G6RRTF3QK7OL7ANCNFSM4IY7DBTQ>
> .
>

justwheel commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+      - 80:80
+      - 443:443
+  db:
+    image: mysql:latest
```
Pinning to a specific version lowers the testing surface by standardizing on a specific version we know works. Since MariaDB 10.1 is available [via the IUS repos](https://ius.io/LifeCycle/#mariadb), I suggest pinning this to `mariadb:10.1`.

justwheel commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+    image: mysql:latest
+    command: mysql --default-authentication-plugin=mysql_native_password
+    volumes:
+      - "./mysql:/var/lib/mysql"
```
Is there a `$TMPDIR` we can use to avoid storing files in the git repo, e.g. `/tmp/grasa_event_locator:/var/lib/mysql`?

Also, it is not obvious, but for systems using SELinux like CentOS 7, you need to [append an SELinux flag](https://www.projectatomic.io/blog/2015/06/using-volumes-with-docker-can-cause-problems-with-selinux/) to the volume string with `:z`.

nlMeminger replied:
using a directory in /tmp isn't a good idea because files in that directory are deleted on reboot.  In order to keep our application data persistent, we should use a non /tmp directory, such as /grasa/db/

justwheel replied:
@nlMeminger I guess it depends whether you want your database to be ephemeral or not :slightly_smiling_face: We can go with this for now, but don't forget the `:z` SELinux volume flag.

nlMeminger replied:
@jwflory works for me

justwheel commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+      - MYSQL_ROOT_PASSWORD=djangoGrasa2019
+      - MYSQL_DATABASE=grasa_event_locator
+      - MYSQL_USER=grasaadmin
+      - MYSQL_PASSWORD=djangoGrasa2019
```
Nice, this is the best place to set these :+1: They can also be overriden on the command line.

justwheel commented on grasa_event_locator/settings/production.py:
```diff
@@ -7,12 +7,13 @@
+        'USER': 'grasaadmin',
+        'PASSWORD': 'djangoGrasa2019',
+        'HOST': 'db',
+        'PORT': '3306',
```
These options should not be hard-coded into the configuration file. These values should be retrieved from environment variables since they are set in the `docker-compose` file (and in future production, they will be set in that environment).

nlMeminger replied:
please see #12

justwheel replied:
Same feedback as [here](https://github.com/jwflory/django-rit-grasa/pull/11/files#r326924238). Could you please configure this as the development database instead in `development.py`?

justwheel reviewed (changes requested):
Hey @nlMeminger, thanks for getting these pushed up! :raised_hands: This looks good. I left some comments with more details in-line. I also pushed a few commits for some quick nitpicks. Please take a look at the feedback today when you can.

justwheel committed:
Use a statically versioned database in docker-compose

This commit changes the database to use the `mariadb:10.1` image currently available for installation via the IUS repositories on CentOS 7. This helps us bring closer parity between local development environment and the final production environment.

justwheel committed:
Dockerfile: Fix flag typo

This commit fixes a hyphen typo in the Dockerfile.

justwheel commented on grasa_event_locator/settings/base.py:
```diff
@@ -27,7 +27,7 @@
 
 # https://docs.djangoproject.com/en/2.2/ref/settings/#allowed-hosts
-ALLOWED_HOSTS = []
+ALLOWED_HOSTS = ["grasa.larrimore.de", "abba.larrimore.de"]
```
These hosts should be set via environment variables passed to the container on start time. I think it is better to leave this blank in order to ship an empty default configuration.

nlMeminger replied:
you're right, which is why I just created #12 but in the meantime, this should stay until we can resolve that issue.

justwheel replied:
@nlMeminger Okay, fair. Instead of putting this in the `base` configuration, could you please add it to the development environment settings in `grasa_event_locator/settings/development.py`? This way it won't be inherited in a production deployment.

nlMeminger replied:
I think for now it should stay until the issue is resolved since this is needed in order to access the site from a url.  This can always be changed at a later time as long as it's noted.

nlMeminger committed:
Update Dockerfile

Fixed white space and postgres and --system flag

nlMeminger committed:
Update docker-compose.yml

nlMeminger committed:
Update development.py

nlMeminger committed:
Update base.py

justwheel committed:
Volumes: Set volume path to cache directory

This commit sets the local volume path to a location in the user's home directory instead of a folder at root. The user running the script may not have a `/grasa` directory at root.

justwheel committed:
Revert changes to production.py

Limit hard-coded options to the development settings file in `development.py`.

justwheel committed:
Update production.py

justwheel committed:
Fix conflicting changes

The `--system` flag was duplicated and the `apt-get` steps required manual intervention. This fixes them and condenses the `RUN` steps to a single layer.

justwheel committed:
Revert addition to gitignore

Since the MySQL database will be stored somewhere else (i.e. not in the git repo), this path is not required in the `gitignore`.

justwheel reviewed (approved):
Changes look good to me. In the future, it would be good to [rebase these commits](https://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/) to fewer commits. I'll do a squash+merge since it's late.
